src/storage-client/src/storage_collections.rs

```diff
@@ -2232,7 +2232,15 @@ where
 
         let gen_upper_future = |id: GlobalId, mut handle: WriteHandle<SourceData, (), T, i64>| {
             let fut = async move {
-                let current_upper = handle.shared_upper();
+                // Note: It is important that we _don't_ use `handle.shared_upper()` here. The
+                // shared upper may have advanced concurrently, in which case we should report that
+                // advancement. However, if we assumed the shared upper as the current upper, we
+                // would be waiting for the next advancement instead. At best this would introduce
+                // frontier reporting delays. At worst the shard could have advanced to the empty
+                // frontier, the `wait_for_upper_past` call would block forever, and we would never
+                // report the empty frontier.
+                let current_upper = handle.upper().clone();
```

Comment:
I'd consider passing this frontier as an argument to the function instead of adding this reasoning here. (Though you'd probably have to initialize it with a call to `handle.upper()` on some paths.) Would make it more explicit that the future will resolve if the frontier advances from where it was at the previous call...

Reply:
Yeah, that's my sense too. `BackgroundCmd::Register` is the only place where we don't already know the previous frontier. I wasn't sure what frontier we should pass in this case, but I guess initializing to `handle.upper()` is safe, since that retains the current behavior.

Reply:
Changed in a fixup commit.